### PR TITLE
Feature/aggregation level

### PR DESCRIPTION
**Pull request by KrzysztofLabus-RDX**

### CODE REVIEW

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs [hunk 1]
```diff
@@ -601,46 +597,125 @@ LIMIT 1
...
+            ? new GatewayModel.EntityFungiblesCursor(resourceLimit).ToCursorString()
+            : null;
+
+        return new GatewayModel.FungibleResourcesCollection(resourcesTotalCount, null, nextCursor, resources.Values.Cast<GatewayModel.FungibleResourcesCollectionItem>().ToList());
+    }
+
+    /// <summary>
+    /// Returns a paginable resource collection with total (aggregated) count per resource.
+    /// </summary>
+    private async Task<GatewayModel.NonFungibleResourcesCollection> GetNonFungiblesSlice(long entityId, int offset, int limit, GatewayModel.LedgerState ledgerState, CancellationToken token)
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:694]: can we rename it to `GetAggregatedNonFungibleSlice` so we could easily distinguish it from `GetNonFungiblesSlicePerVault`?

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs [hunk 2]
```diff
@@ -527,8 +535,10 @@ LIMIT 1
         return result;
     }
 
-    private async Task<GatewayModel.FungibleResourcesCollection> GetFungiblesSlice(long entityId, int offset, int limit, GatewayModel.LedgerState ledgerState,
-        CancellationToken token)
+    /// <summary>
+    /// Returns a paginable resource collection with total (aggregated) balance per resource.
+    /// </summary>
+    private async Task<GatewayModel.FungibleResourcesCollection> GetFungiblesSlice(long entityId, int offset, int limit, GatewayModel.LedgerState ledgerState, CancellationToken token)
```
Comment (PawelPawelec-RDX) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:541]: can we rename it to `GetAggregatedFungibleSlice` so we could easily distinguish it from `GetFungiblesSlicePerVault`?

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 3]
```diff
@@ -1224,7 +1306,7 @@ components:
           required:
             - address
             - fungible_resources
-            - non_fungible_resources
+            - non_fungible_resources\
```
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:1356]: Fixed

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs [hunk 4]
```diff
@@ -121,13 +123,15 @@ public EntityStateQuerier(INetworkConfigurationProvider networkConfigurationProv
...
-    public async Task<GatewayModel.EntityResourcesResponse> EntityResourcesSnapshot(GlobalAddress address, GatewayModel.LedgerState ledgerState, CancellationToken token = default)
+    public async Task<GatewayModel.EntityResourcesResponse> EntityResourcesSnapshot(GlobalAddress address, bool aggregatePerVault, GatewayModel.LedgerState ledgerState, CancellationToken token = default)
     {
         var entity = await GetEntity<ComponentEntity>(address, ledgerState, token);
 
         // TODO ideally we'd like to run those as either single query or separate ones but without await between them
 
-        var fungibles = await GetFungiblesSlice(entity.Id, 0, DefaultResourceLimit, ledgerState, token);
+        var fungibles = aggregatePerVault
+            ? await GetFungiblesSlicePerVault(entity.Id, DefaultResourceLimit, DefaultResourceLimit, ledgerState, token)
```
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:135]: Please note we're doing `GetEntity<ComponentEntity>` so no virtual entities will be returned but indeed we might need to rework logic slightly for future `/state/entity/details` endpoint and avoid those pseudo-entities with `Id=0` altogether.

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml [hunk 5]
```diff
@@ -1215,6 +1292,11 @@ components:
           properties:
             address:
               $ref: "#/components/schemas/EntityAddress"
+            aggregation_level:
```
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:1342]: This is not 100% accurate. Default value is "missing" as this is optional field and how to treat missing value might be quite endpoint-specific.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs [hunk 6]
```diff
@@ -601,6 +595,100 @@ LIMIT 1
...
+            parameters: new
+            {
+                stateVersion = ledgerState.StateVersion,
+                entityId = entityId,
+                resourceLimit = resourceLimit,
+                vaultLimit = vaultLimit,
+            },
+            cancellationToken: token);
+
+        long? resourcesTotalCount = 0;
```
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:655]: It has been configured a nullable long to match first parameter of `GatewayModel.FungibleResourcesCollection(totalCount` and similar. But with current implementation it clearly isn't ever going to be null. Fixed.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs (hunk 1 above)
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:694]: Renamed to `*PerResource` to match `*PerVault` scheme.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:135]: `VirtualAccountComponentEntity` inherits from `AccountComponentEntity` which inherits from `ComponentEntity`.

I tried to call that exact endpoint with virtual component address `account_tdx_22_1qusjhfslr6v2s7xe33vzeyjurnq8aqptyf5z7jadm39s307qh3` and it returned me virtual component.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs (hunk 2 above)
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs]: Renamed to *PerResource to match *PerVault scheme.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:135]: Ah, right! Obviously you're right. Still this does not hurt us and I guess earlier comment still stays relevant:

> [...] but indeed we might need to rework logic slightly for future /state/entity/details endpoint and avoid those pseudo-entities with Id=0 altogether.

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:1342]: that's right. Do we want to mention that in endpoint definition then?

File: src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:135]: would you mind dropping todo next to these virtual components code?:P it's worth leaving trace of that conversation somewhere.

[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/Services/EntityStateQuerier.cs:135]: Added

File: src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (KrzysztofLabus-RDX) [reply src/RadixDlt.NetworkGateway.GatewayApi/gateway-api-schema.yaml:1342]: We'll do it in separate PR related to `/state/entity/details` endpoint then.

**Review by PawelPawelec-RDX — APPROVED** [review]

**Pull request merged**